File: examples/tailcat-e2b/js/src/sandboxToSandbox.ts
/**
 * Two E2B sandboxes exchange encrypted data through Tailcat's DERP relay.
 *
 * The producer sandbox generates a dataset and serves a read-write directory.
 * The consumer sandbox pulls the dataset over Tailcat, analyzes it, and pushes
 * a JSON report back into the producer's directory. The orchestrator only passes
 * the producer's address to the consumer; file bytes do not pass through it.
 */
import "dotenv/config";
import { setTimeout as delay } from "node:timers/promises";
import {
  computeSandboxMd5,
  createSandbox,
  createSandboxCommandRunner,
  describeConnection,
  formatMegabitsPerSecond,
  requireSuccessfulCommand,
  runSandboxCommand,
  secondsSince,
  startSandboxTailcatServer,
  waitUntilReachable,
} from "./tailcatRuntime";

const FILE_SIZE_MIB = Number(process.env.DEMO_FILE_SIZE_MIB ?? 10);
const FILE_SIZE_BYTES = FILE_SIZE_MIB * 1024 * 1024;

const producerSandbox = await createSandbox();
const consumerSandbox = await createSandbox();
try {
  // Producer: generate a dataset and share its directory read-write.
  await producerSandbox.commands.run(
    `mkdir -p /home/user/share && head -c ${FILE_SIZE_BYTES} /dev/urandom > /home/user/share/dataset.bin`,
  );
  const sharedDirectoryServer = await startSandboxTailcatServer(
    producerSandbox,
    "serve --files=/home/user/share:rw files",
    "share",
  );
  try {
    // Consumer: connect using only the address string.
    const runInConsumer = createSandboxCommandRunner(consumerSandbox, 300_000);
    await waitUntilReachable(runInConsumer, sharedDirectoryServer.address);
    console.log(
      "[consumer] " +
        (await describeConnection(
          runInConsumer,
          sharedDirectoryServer.address,
        )),
    );

    const directoryListing = await runInConsumer(
      `tailcat ls -l ${sharedDirectoryServer.address}`,
    );
    requireSuccessfulCommand(directoryListing);
    console.log(
      "[consumer] tailcat ls -l <producer>:\n" + directoryListing.output.trim(),
    );

    let startedAt = Date.now();
    requireSuccessfulCommand(
      await runInConsumer(
        `tailcat cp ${sharedDirectoryServer.address}:dataset.bin /home/user/dataset.bin`,
      ),
    );
    let elapsedSeconds = secondsSince(startedAt);
    const checksumMatches =
      (await computeSandboxMd5(consumerSandbox, "/home/user/dataset.bin")) ===
      (await computeSandboxMd5(
        producerSandbox,
        "/home/user/share/dataset.bin",
      ));
    if (!checksumMatches) {
      throw new Error(
        "the dataset checksum differs between the producer and consumer",
      );
    }
    console.log(
      `[producer -> consumer] ${FILE_SIZE_MIB} MiB in ${elapsedSeconds.toFixed(
        1,
      )}s ` +
        `(${formatMegabitsPerSecond(FILE_SIZE_BYTES, elapsedSeconds)}), md5 ok`,
    );

    // Consumer: analyze the dataset, then push a useful result back to the producer.
    const analysis = await runInConsumer(
      `checksum=$(sha256sum /home/user/dataset.bin | cut -d' ' -f1); ` +
        `printf '{"bytesProcessed":${FILE_SIZE_BYTES},"status":"complete","generatedBy":"worker-sandbox","sha256":"%s"}\n' ` +
        `"$checksum" > /home/user/analysis-report.json`,
    );
    requireSuccessfulCommand(analysis);
    startedAt = Date.now();
    requireSuccessfulCommand(
      await runInConsumer(
        `tailcat cp /home/user/analysis-report.json ${sharedDirectoryServer.address}:`,
      ),
    );
    elapsedSeconds = secondsSince(startedAt);
    const processedResult = await runSandboxCommand(
      producerSandbox,
      "python3 -m json.tool /home/user/share/analysis-report.json",
    );
    requireSuccessfulCommand(processedResult);
    console.log(
      `[consumer -> producer] analysis-report.json returned in ${elapsedSeconds.toFixed(
        1,
      )}s`,
    );
    console.log("\nFinal report:\n" + processedResult.output.trim());

    // Streaming variant: bare `tailcat` accepts one connection and copies it to
    // stdout, so the producer receives whatever the consumer pipes in.
    const streamReceiver = await startSandboxTailcatServer(
      producerSandbox,
      "",
      "stream",
      {
        stdoutFile: "/home/user/share/stream.log",
      },
    );
    try {
      await waitUntilReachable(runInConsumer, streamReceiver.address);
      requireSuccessfulCommand(
        await runInConsumer(
          `seq 1 5 | sed 's/^/log line /' | tailcat ${streamReceiver.address}`,
        ),
      );
      await delay(1000);
      const receivedStream = await runSandboxCommand(
        producerSandbox,
        "cat /home/user/share/stream.log",
      );
      requireSuccessfulCommand(receivedStream);
      console.log(
        "[consumer -> producer stream] producer received:\n" +
          receivedStream.output.trim(),
      );
    } finally {
      await streamReceiver.stop();
    }
  } finally {
    await sharedDirectoryServer.stop();
  }
} finally {
  await producerSandbox.kill();
  await consumerSandbox.kill();
}

File: examples/tailcat-e2b/python/tailcat_e2b/sandbox_to_sandbox.py
"""Two E2B sandboxes exchange encrypted data through Tailcat's DERP relay.

The producer sandbox generates a dataset and serves a read-write directory.
The consumer sandbox pulls the dataset over Tailcat, analyzes it, and pushes
a JSON report back into the producer's directory. The orchestrator only passes
the producer's address to the consumer; file bytes do not pass through it.
"""

import os
import sys
import time

from dotenv import load_dotenv
from e2b import Sandbox

from .e2b_sandbox import (
    assert_command_succeeded,
    create_sandbox,
    run_sandbox_command,
    sandbox_command_runner,
)
from .tailcat_server import (
    start_sandbox_tailcat_server,
    wait_until_reachable,
)

FILE_SIZE_MIB = int(os.environ.get("DEMO_FILE_SIZE_MIB", "10"))
FILE_SIZE_BYTES = FILE_SIZE_MIB * 1024 * 1024


def compute_sandbox_md5(sandbox: Sandbox, path: str) -> str:
    result = run_sandbox_command(sandbox, f"md5sum {path}")
    assert_command_succeeded(result)
    return result.output.split()[0]


def transfer_speed(size_bytes: int, seconds: float) -> str:
    return f"{size_bytes * 8 / seconds / 1e6:.0f} Mbit/s"


def main() -> None:
    sys.stdout.reconfigure(line_buffering=True)
    load_dotenv()
    producer_sandbox = create_sandbox()
    consumer_sandbox = create_sandbox()
    try:
        # Producer: generate a dataset and share its directory read-write.
        producer_sandbox.commands.run(
            f"mkdir -p /home/user/share && head -c {FILE_SIZE_BYTES} /dev/urandom > /home/user/share/dataset.bin",
        )
        shared_directory_server = start_sandbox_tailcat_server(
            producer_sandbox, "serve --files=/home/user/share:rw files", "share"
        )
        try:

            # Consumer: connect using only the address string.
            run_in_consumer = sandbox_command_runner(consumer_sandbox, timeout_seconds=300)
            connection = wait_until_reachable(run_in_consumer, shared_directory_server.address)
            print(f"[consumer] {connection}")

            directory_listing = run_in_consumer(f"tailcat ls -l {shared_directory_server.address}")
            assert_command_succeeded(directory_listing)
            print("[consumer] tailcat ls -l <producer>:\n" + directory_listing.output.strip())

            started_at = time.time()
            assert_command_succeeded(
                run_in_consumer(f"tailcat cp {shared_directory_server.address}:dataset.bin /home/user/dataset.bin")
            )
            elapsed_seconds = time.time() - started_at
            checksum_matches = compute_sandbox_md5(consumer_sandbox, "/home/user/dataset.bin") == compute_sandbox_md5(
                producer_sandbox, "/home/user/share/dataset.bin"
            )
            if not checksum_matches:
                raise RuntimeError("the dataset checksum differs between the producer and consumer")
            print(
                f"[producer -> consumer] {FILE_SIZE_MIB} MiB in {elapsed_seconds:.1f}s "
                f"({transfer_speed(FILE_SIZE_BYTES, elapsed_seconds)}), md5 ok"
            )

            # Consumer: analyze the dataset, then push a useful result back to the producer.
            analysis = run_in_consumer(
                "checksum=$(sha256sum /home/user/dataset.bin | cut -d' ' -f1); "
                f"printf '{{\"bytesProcessed\":{FILE_SIZE_BYTES},\"status\":\"complete\","
                "\"generatedBy\":\"worker-sandbox\",\"sha256\":\"%s\"}\\n' "
                '"$checksum" > /home/user/analysis-report.json'
            )
            assert_command_succeeded(analysis)
            started_at = time.time()
            assert_command_succeeded(
                run_in_consumer(f"tailcat cp /home/user/analysis-report.json {shared_directory_server.address}:")
            )
            elapsed_seconds = time.time() - started_at
            processed_result = run_sandbox_command(
                producer_sandbox, "python3 -m json.tool /home/user/share/analysis-report.json"
            )
            assert_command_succeeded(processed_result)
            print(f"[consumer -> producer] analysis-report.json returned in {elapsed_seconds:.1f}s")
            print("\nFinal report:\n" + processed_result.output.strip())

            # Streaming variant: bare `tailcat` accepts one connection and copies it to
            # stdout, so the producer receives whatever the consumer pipes in.
            stream_receiver = start_sandbox_tailcat_server(
                producer_sandbox, "", "stream", stdout_file="/home/user/share/stream.log"
            )
            try:
                wait_until_reachable(run_in_consumer, stream_receiver.address)
                assert_command_succeeded(
                    run_in_consumer(f"seq 1 5 | sed 's/^/log line /' | tailcat {stream_receiver.address}")
                )
                time.sleep(1)
                received_stream = run_sandbox_command(producer_sandbox, "cat /home/user/share/stream.log")
                assert_command_succeeded(received_stream)
                print("[consumer -> producer stream] producer received:\n" + received_stream.output.strip())
            finally:
                stream_receiver.stop()
        finally:
            shared_directory_server.stop()
    finally:
        producer_sandbox.kill()
        consumer_sandbox.kill()


if __name__ == "__main__":
    main()
